cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -1480,89 +1482,117 @@ func maybeMountSave(disk disks.Disk, rootdir string, encrypted bool, mountOpts *
 	return true, nil
 }
 
-func generateMountsModeRun(mst *initramfsMountsState) error {
+func generateMountsModeRun(mst *initramfsMountsState) (string, error) {
 	// 1. mount ubuntu-boot
 	if err := mountNonDataPartitionMatchingKernelDisk(boot.InitramfsUbuntuBootDir, "ubuntu-boot"); err != nil {
-		return err
+		return "", err
 	}
 
 	// get the disk that we mounted the ubuntu-boot partition from as a
 	// reference point for future mounts
 	disk, err := disks.DiskFromMountPoint(boot.InitramfsUbuntuBootDir, nil)
 	if err != nil {
-		return err
+		return "", err
+	}
+
+	// 1.1. measure model
+	err = stampedAction("run-model-measured", func() error {
+		return secbootMeasureSnapModelWhenPossible(mst.UnverifiedBootModel)
+	})
+	if err != nil {
+		return "", err
+	}
+
+	// The model is now measured, use it to check if this is a classic install
+	model, err := mst.UnverifiedBootModel()
+	if err != nil {
+		return "", err
+	}
+	isClassic := model.Classic()
+	var rootfsDir string
+	if isClassic {
+		logger.Noticef("generating mounts for classic system, run mode")
+		rootfsDir = boot.InitramfsDataDir
+	} else {
+		logger.Noticef("generating mounts for Ubuntu Core system, run mode")
+		rootfsDir = boot.InitramfsWritableDir
 	}
 
-	// 2. mount ubuntu-seed
+	// 2. mount ubuntu-seed (optional for classic)
+	systemdOpts := &systemdMountOptions{
+		NeedsFsck: true,
+		Private:   true,
+	}
 	// use the disk we mounted ubuntu-boot from as a reference to find
 	// ubuntu-seed and mount it
 	partUUID, err := disk.FindMatchingPartitionUUIDWithFsLabel("ubuntu-seed")
 	if err != nil {
-		return err
+		if isClassic {
+			// If there is no ubuntu-seed on classic, that's fine
+			if _, ok := err.(disks.PartitionNotFoundError); !ok {
+				return "", err
+			}
```

Comment:
I don't see code below to check that the expectation from the gadget about this is met

Reply:
Not sure if I understand this comment, the gadget that we mount comes from the ubuntu-boot partition, not from ubuntu-seed.

Reply:
The gadget that we mount should come from the data partition, no?  What I'm seeing I don't see a check that we don't end up not having found a ubuntu-seed while our gadget volumes tell there should be one

Reply:
Good point, I've added checks for this